File: source/com/barrybecker4/optimization/optimizees/OptimizeeProblem.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.optimization.optimizees;

import com.barrybecker4.optimization.Optimizee;
import com.barrybecker4.optimization.parameter.ParameterArray;

/**
 * Abstract base class for optimizer test problems.
 *
 * @author Barry Becker
 */
public abstract class OptimizeeProblem implements Optimizee {

    /**
     * @return  the exact solution for this problem.
     */
    public abstract ParameterArray getExactSolution();

    /**
     * @return  the exact solution for this problem.
     */
    public abstract ParameterArray getInitialGuess();

    /**
     *
     * @param sol solution
     * @return distance from the exact solution as the error.
     */
    public double getError(ParameterArray sol) {
        return 100.0 * sol.distance(getExactSolution()) / getFitnessRange();
    }

    @Override
    public double getOptimalFitness() {
        return 0;
    }

    /**
     *
     * @return  approximate range of fitness values (usually 0 to this number).
     */
    public abstract double getFitnessRange();

}

File: scala-source/com/barrybecker4/optimization/optimizee/optimizees/OptimizeeProblem.scala
// Copyright by Barry G. Becker, 2000-2018. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.optimization.optimizee.optimizees

import com.barrybecker4.optimization.optimizee.Optimizee
import com.barrybecker4.optimization.parameter.{ParameterArray, ParameterArrayWithFitness}


object OptimizeeProblem {

  def showSolution(problem: OptimizeeProblem[ParameterArray], solution: ParameterArrayWithFitness[ParameterArray]): Unit = {
    println("\n************************************************************************")
    println("The solution to the " + problem.getName + " test problem is :\n" + solution)
    println("Which evaluates to: " + solution.fitness)
    println("We expected to get exactly " + problem.getExactSolution)
  }
}

/**
  * Abstract base class for optimizer test problems.
  * @author Barry Becker
  */
abstract class OptimizeeProblem[P <: ParameterArray] extends Optimizee[P] {

  /** @return the exact solution for this problem. */
  def getExactSolution: ParameterArrayWithFitness[P]

  def getInitialGuess: P

  /** @return distance from the exact solution as the error. */
  def getError(solution: ParameterArrayWithFitness[P]): Double =
    100.0 * (solution.fitness - getOptimalFitness) / getFitnessRange

  override def getOptimalFitness = 0
  override def compareFitness(a: P, b: P) =
    throw new UnsupportedOperationException("compareFitness not supported for " + this.getClass.getName)
  override def toString: String = getName

  /** @return approximate range of fitness values (usually 0 to this number). */
  def getFitnessRange: Double
}